README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# curvematching

<!-- badges: start -->

[![R-CMD-check](https://github.com/growthcharts/curvematching/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/growthcharts/curvematching/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of `curvematching` is to …

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("growthcharts/curvematching")
```


NEWS.md
# curvematching 0.14.0

* `tidyselect` 1.2.0 grammar for the `dplyr::select()` function now forbids `.data$varname` arguments. Update various function to remove deprecation messages and errors.
* Uses the more lenient `any_of()` to prevent hard errors.


# curvematching 0.13.1

* Removes superfluous mutate_() imports
* Removes superfluous LazyData in DESCRIPTION

# curvematching 0.13.0

* Adds GHA `R-CMD-check`
* Merges the `bdsreader` branch so as to remove `minihealth`
* Relocates repo to the `growthchart` organisation

# curvematching 0.12.0

* Adds blended distance matching function `match_bmd()`, which generalises the distance metric used for matching

# curvematching 0.11.0

* Replaces `substitute()` by tidyeval `enexpr()`

# curvematching 0.10.0

* Removes the !!! expression, deprecated in `R 4.0.0`
* Updates a test after requiring `donorloader 0.12.0`

# curvematching 0.9.1

* Update to R 4.0.0

# curvematching 0.9.0

* Repairs a bug in `match_pmm()` that could produce big blob of matches

# curvematching 0.8.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library("testthat")
library("curvematching")

test_check("curvematching")


tests/testthat/test_calculate_matches.R
context("calculate_matches")

set.seed(234)
data <- datasets::ChickWeight

test_that("returns proper length", {
  expect_equal(length(calculate_matches(data, Time == 0, y_name = "sasa")), 0)
  expect_equal(length(calculate_matches(data, Time == 0 & Chick == 48, y_name = "weight")), 1)
  expect_equal(length(calculate_matches(data, Time == 0 & Diet == 2, y_name = "weight")), 10)
})

test_that("warns if no targets are found", {
  expect_warning(calculate_matches(data, Time == 0 & Chick == 2 & Diet == 2,
    y_name = "weight"
  ))
})

test_that("warns about rank-deficient fit", {
  expect_warning(calculate_matches(data, Chick == 48 & Time == 0,
    y_name = "weight",
    x_name = c("Chick", "Diet"),
    e_name = c("Time")
  ))
})


context("extract_matches()")
set.seed(234)
z0 <- calculate_matches(data, Time == 0, y_name = "sasa")
z10 <- calculate_matches(data, Time == 0 & Chick == 48,
  y_name = c("weight", "Time"), t_name = "Diet"
)

test_that("return integer(0) if there are matches", {
  expect_equal(length(extract_matches(z0)), 0L)
})

test_that("returns vector of 10 row indices", {
  expect_equal(
    extract_matches(z10, y_name = "Time", t_name = "Diet", c_name = "3"),
    c(342, 369, 375, 411, 414, 423, 428, 438, 457, 459)
  )
})

z <- calculate_matches(data, Chick == 2, y_name = c("weight", "Time"))
test_that("returns vector of 10 row indices", {
  expect_equal(
    extract_matches(z, i_name = "23", y_name = "Time"),
    c(128, 167, 243, 244, 245, 299, 407, 420, 472, 494)
  )
})

#
# z <- calculate_matches(ChickWeight, list(Chick = 2), y_name = c("weight", "Time"), t_name = "Diet")
# extract_matches(z)
extract_matches(z, i_name = "23", y_name = "Time")
extract_matches(z, i_name = "23", y_name = "Time", t_name = "Diet", c_name = "3")
#
# z <- calculate_matches(ChickWeight, list(Chick = 2), y_name = c("weight", "Time"), t_name = "Time")
# extract_matches(z)
# extract_matches(z, i_name = "23", y_name = "Time")
extract_matches(z, i_name = "23", y_name = "Time", t_name = "Diet", c_name = "2")


tests/testthat/test_calculate_matches2.R
context("calculate_matches2")

set.seed(234)
data <- datasets::ChickWeight

test_that("skips over crappy y_name", {
  expect_silent(calculate_matches2(data, data[data$Time == 0, ], y_name = "sasa"))
})

test_that("returns proper length", {
  expect_equal(length(calculate_matches2(data, data[data$Time == 0 & data$Chick == 48, ], y_name = "weight")), 1)
  expect_equal(length(calculate_matches2(data, data[data$Time == 0 & data$Diet == 2, ], y_name = "weight")), 10)
})

test_that("warns if no targets are found", {
  expect_warning(calculate_matches2(data, data[data$Time == 0 & data$Chick == 2 & data$Diet == 2, ], y_name = "weight"))
})

test_that("warns about rank-deficient fit", {
  expect_warning(calculate_matches2(data, data[data$Chick == 48 & data$Time == 0, ],
    y_name = "weight",
    x_name = c("Chick", "Diet"),
    e_name = c("Time")
  ))
})

# select some row
row <- 200
m <- calculate_matches2(data, data[row, ],
  y_name = "weight",
  x_name = c("Chick", "Time"),
  k = 1
)
test_that("closest predictor is target row", {
  expect_equal(extract_matches(m), row)
})


context("extract_matches()")
set.seed(234)
z0 <- calculate_matches2(1)

test_that("return integer(0) if there are matches", {
  expect_equal(length(extract_matches(z0)), 0L)
})

# z2 <- calculate_matches2(data, data[data$Time == 0 & data$Chick == 48, ],
#                          y_name = c("weight", "Time"), t_name = "Diet")
# m2 <- extract_matches(z2, y_name = "Time", t_name = "Diet", c_name = "3")
# test_that("returns vector of 10 row indices", {
#   expect_equal(m2,
#                c(342, 369, 375, 411, 414, 423, 428, 438, 457, 459))
# })
# data[m2, ]

# z <- calculate_matches2(data, data[data$Chick == 2, ], y_name = c("weight", "Time"))
# test_that("returns vector of 10 row indices", {
#   expect_equal(extract_matches(z, i_name = "23", y_name = "Time"),
#                c(128, 167, 243, 244, 245, 299, 407, 420, 472, 494))
# })
#
# #
# # z <- calculate_matches2(ChickWeight, list(Chick = 2), y_name = c("weight", "Time"), t_name = "Diet")
# # extract_matches(z)
# extract_matches(z, i_name = "23", y_name = "Time")
# extract_matches(z, i_name = "23", y_name = "Time", t_name = "Diet", c_name = "3")
# #
# # z <- calculate_matches2(ChickWeight, list(Chick = 2), y_name = c("weight", "Time"), t_name = "Time")
# # extract_matches(z)
# # extract_matches(z, i_name = "23", y_name = "Time")
# extract_matches(z, i_name = "23", y_name = "Time", t_name = "Diet", c_name = "2")



# equivalence between calculate_matches() and calculate_matches2()
# Note: not generic, holds for k = 9. May be different for other k
m <- calculate_matches(data, Time == 0 & Chick == 48, y_name = "weight", x_name = c("Time", "Diet"), k = 9)
m2 <- calculate_matches2(data, data[543, ], subset = !rownames(data) %in% "543", y_name = "weight", x_name = c("Time", "Diet"), k = 9)

test_that("calculate_matches() and calculate_matches2() return same matches", {
  expect_equal(extract_matches(m), extract_matches(m2))
})


R/extract_matches.R
#' Extracts the calcated matches for indexing
#'
#' @param matches An object created by `calculate_matches()`. Such an
#' object has class `"match_list"`.
#' @param i_name The name of the indvidual for which matches are wanted.
#' If unspecified, it takes the first individual.
#' @param y_name The name of the outcome variable. If unspecified, it takes the
#' first outcome.
#' @param t_name A character vector containing the name of the treatment
#' variables.
#' @param c_name The name of the category level within the treatment variable.
#' @return The row numbers in
#' `data` corresponding to the matched children. The length
#' of the list will be always equal to `m` if `replace == TRUE`,
#' but may be shorter if `replace == FALSE` if the donors are exhausted. The
#' length is zero if no matches can be found.
#' @author Stef van Buuren 2016
#' @references
#' van Buuren, S. (2014). *Curve matching: A data-driven technique to
#' improve individual prediction of childhood growth*. Annals of Nutrition &
#' Metabolism, 65(3), 227-233.
#' van Buuren, S. (2012). *Flexible imputation of missing data*.
#' Boca Raton, FL: Chapman & Hall/CRC.
#' @export
extract_matches <- function(matches,
                            i_name = names(matches)[[1]],
                            y_name = names(matches[[i_name]])[[1]],
                            t_name = character(0),
                            c_name = character(0)) {
  if (!inherits(matches, "match_list")) stop("Argument `matches` not of class `match_list`.")

  if (length(matches) == 0) {
    return(integer(0))
  }

  found <- i_name %in% names(matches)
  if (any(!found)) {
    warning("i_name elements ", i_name[!found], " not found.")
    return(integer(0))
  }

  found <- y_name %in% names(matches[[i_name]])
  if (any(!found)) {
    warning("y_name elements", y_name[!found], " not found.")
    return(integer(0))
  }

  # if there is no .by variable, then simply ignore
  # any treatment parameters and return first row
  tab <- matches[[i_name]][[y_name]]
  has_by <- tab[1, ".by"]
  if (!has_by) {
    return(as.vector(unlist(tab[1, ".row"])))
  }

  # handle treatment selection: t_name and c_name
  if (length(t_name) == 0) {
    tn <- names(tab)
    t_name <- tn[!tn %in% c(".by", ".row")][1]
  }
  if (!t_name %in% names(tab)) {
    warning("Name (t) ", t_name, " not found.")
    return(integer(0))
  }
  if (length(c_name) == 0) c_name <- unlist(tab[1, t_name])
  if (!c_name %in% unlist(tab[, t_name])) {
    warning("Name (c) ", c_name, " not found.")
    return(integer(0))
  }
  as.vector(unlist(tab[unlist(tab[, t_name]) == c_name, ".row"]))
}


R/calculate_matches2.R
#' Calculates matches for one or more children by blended distance matching
#'
#' Curve matching is a technology that aims to predict individual growth curves.
#' The method finds persons similar to the target person, and learn the possible
#' future course of growth from the realized curves of the matched individuals.
#'
#' The function finds `k` matches for an individual in the same data set by
#' means of stratified predictive mean matching or by nearest neighbour matching.
#'
#' @param data A `data.frame` or `tbl_df` with donor data. One row is
#' one potential donor.
#' @param newdata A `data.frame` or `tbl_df` with data of children for
#'  which we seek matches. Every row corresponds to one child.
#' @param y_name A character vector containing the names of the dependent
#'   variables in `data`.
#' @param x_name A character vector containing the names of predictive variables
#'   in `data` to will go into the linear part of the model.
#' @param e_name A character vector containing the names of the variables for
#'   which the match should be exact.
#' @param t_name A character vector containing the names of the treatment
#'   variables in `data`. The current function will only fit the model to
#'   only the first element of `t_name`.
#' @param subset Logical expression defining the set of rows taken from
#'   `data`. This subset is selected before any other calculations are
#'   made, and this can be used to trim down the size of the data in which
#'   matches are defined and sought.
#' @param k Requested number of matches. The default is `k = 10`.
#' @param replace A logical that indicates whether to match with or without
#'   replacement. The default is `FALSE`.
#' @param blend An integer value between 0 and 1 that indicates the blend
#'   between predictive mean matching with replacement (`1`) and euclidian
#'   distance matching (`0`). The default is `1`.
#' @param kappa A numeric value that serves as the sensitivity parameter for the
#'   inverse distance weighting. Used when drawing with replacement. The default
#'   is `3`.
#' @param break_ties A logical indicating whether ties should broken randomly.
#'   The default (`TRUE`) breaks ties randomly.
#' @param verbose A logical indicating whether diagnostic information should be
#'   printed.
#' @param \dots Arguments passed down to `match_pmm()`.
#' @return An object of class `match_list` which can be post-processed by
#'   the `extract_matches` function to extract the row numbers in
#'   `data` of the matched children. The length of the list will be always
#'   equal to `m` if `replace == TRUE`, but may be shorter if
#'   `replace == FALSE` if the donors are exhausted. The length is zero if
#'   no matches can be found.
#' @author Stef van Buuren 2021
#' @details
#'   The procedure search for matches in `data` for each row
#'   in `newdata`.
#'   Note that if `x_name` contains one or more factors, then it is
#'   possible that the factor level of the target case is unique among all
#'   potential donors. In that case, the model can still be fit, but prediction
#'   will fail, and hence no matches will be found.
#'
#'   If `break_ties` is `FALSE`, the function returns the first `nmatch`
#'   matches as they appear in the order of `data`. This method overuses
#'   the first part of the data if there are ties, and hence may underestimate
#'   variability. The default option is to break ties randomly.
#'
#' @references van Buuren, S. (2014). *Curve matching: A data-driven
#'   technique to improve individual prediction of childhood growth*. Annals of
#'   Nutrition & Metabolism, 65(3), 227-233. van Buuren, S. (2012).
#'   *Flexible imputation of missing data*. Boca Raton, FL: Chapman &
#'   Hall/CRC.
#' @examples
#' data <- datasets::ChickWeight
#' data[1, ]
#'
#' # find matches for observation in row 1
#' m1 <- calculate_matches2(data, data[1, ],
#'   subset = !rownames(data) %in% "1",
#'   y_name = "weight", x_name = c("Time", "Diet")
#' )
#'
#' # data of matched cases (may vary because of tie breaking)
#' data[extract_matches(m1), ]
#'
#' # without tie breaking, we pick the earlier rows (not recommended)
#' m2 <- calculate_matches2(data, data[1, ],
#'   subset = !rownames(data) %in% "1",
#'   y_name = "weight", x_name = c("Time", "Diet"), break_ties = FALSE
#' )
#' data[extract_matches(m2), ]
#' @export
calculate_matches2 <- function(data,
                               newdata,
                               y_name = character(0L),
                               x_name = character(0L),
                               e_name = character(0L),
                               t_name = character(0L),
                               subset = TRUE,
                               k = 10L,
                               replace = FALSE,
                               blend = 1.0,
                               break_ties = TRUE,
                               kappa = 3,
                               verbose = TRUE, ...) {
  equals_all <- function(x) {
    if (is.null(names(x)) | !length(x)) {
      return(character(0))
    }
    cv <- vector("character", length(x))
    for (j in 1:length(x)) {
      xj <- x[[j]]
      if (!is.numeric(xj)) {
        cv[j] <- paste0(names(x)[j], ' == "', as.character(xj), '"')
      } else {
        cv[j] <- paste0(names(x)[j], " == ", xj)
      }
    }
    paste0(cv, collapse = " & ")
  }

  # validity checks
  if (!is.data.frame(data)) {
    return(no_match())
  }
  if (!nrow(newdata) || !is.data.frame(newdata)) {
    if (verbose) warning("No rows in newdata.")
    return(no_match())
  }
  if (k <= 0 || !length(y_name) || !hasName(data, y_name)) {
    return(no_match())
  }

  # model variables
  vars <- intersect(
    unique(c(y_name, x_name, e_name, t_name)),
    names(data)
  )

  rvars <- c(".row", vars)

  # preserve donor row number before subset
  data <- data %>%
    mutate(.row = 1L:n()) %>%
    filter({{ subset }}) %>%
    select(any_of(rvars))

  newdata <- newdata %>%
    mutate(.row = 1L:n()) %>%
    select(any_of(rvars))

  # loop over target children
  l1 <- vector("list", nrow(newdata))
  names(l1) <- as.vector(unlist(select(newdata, ".row")))
  for (i in 1L:nrow(newdata)) {
    # define active case
    active <- slice(newdata, i)

    # trim candidate set by requiring exact matches on
    # variables listed in `e_name`
    trimmed <- select(active, !!e_name)
    cond <- equals_all(trimmed)
    if (length(cond)) {
      expr <- parse(text = cond)
      data <- data %>%
        mutate(candidate = eval(expr))
    } else {
      data <- data %>%
        mutate(candidate = TRUE)
    }

    # loop over outcome names
    ny <- length(y_name)
    l2 <- vector("list", ny)
    names(l2) <- y_name
    for (iy in 1:ny) {
      yvar <- y_name[iy]

      # extract subset of candidates
      xy <- filter(data, .data$candidate)

      # split by treatment variables
      if (length(t_name) > 0) {
        # duplicate one tgt row for each treatment level
        # FIXME: as.name will only split on first variable name
        t_name <- t_name[[1]]
        # t_name_list <- as.list(t_name)
        t_unique <- unique(xy[, t_name])
        mutate_call <- quo(!!sym(t_name))
        augment <- slice(active, rep(1:n(), each = length(t_unique)))
        augment[, t_name] <- t_unique

        # loop over all treatments
        matched <- vector("list", length(t_unique))
        for (c in 1:nrow(augment)) {
          active_trt <- augment[c, ]
          trt <- t_unique[c]
          data_trt <- filter(data, !!mutate_call == !!trt)
          matched[[c]] <- match_bdm2(
            data = data_trt,
            active = active_trt,
            y_name = yvar, x_name = x_name, k = k, replace = replace,
            blend = blend, break_ties = break_ties, kappa = kappa, ...
          )
        }
        # store
        matched <- augment %>%
          bind_rows(filter(xy, .data$candidate)) %>%
          group_by(!!mutate_call) %>%
          summarise(.row = !!matched) %>%
          mutate(.by = TRUE)

        # this code isn't yet right, do not use t_name for now (SvB 22/3/2021)
        # matched <- augment %>%
        #   bind_rows(filter(xy, .data$candidate))
        # matched <- matched %>%
        #   group_by(!!mutate_call) %>%
        #   do(.row = match_bdm2(., active = active,
        #     y_name = yvar, x_name = x_name, k = k, replace = replace,
        #     blend = blend, break_ties = break_ties, kappa = kappa, ...
        #   )) %>%
        #   mutate(.by = TRUE)
      } else {
        row <- match_bdm2(xy,
          active = active,
          y_name = yvar, x_name = x_name, k = k, replace = replace,
          blend = blend, break_ties = break_ties, kappa = kappa, ...
        )
        matched <- tibble(.by = FALSE, .row = list(row))
      }
      l2[[iy]] <- matched
    }
    l1[[i]] <- l2
  }
  # names(l1) <- child_names
  class(l1) <- "match_list"
  l1
}

# Blended distance matching
match_bdm2 <- function(data, active, y_name, x_name, k, replace = TRUE, blend = 1,
                       break_ties = TRUE, exclude_NA = FALSE, kappa = 3, ...) {
  if (nrow(data) <= 0L) {
    return(no_match())
  }
  if (nrow(active) != 1L) stop("Argument active not one row.")
  if (blend > 1 | blend < 0) stop("blend needs to be a value between 0 and 1.")

  # keep only independent variables taking at least two values
  # note: apparently, next statement cannot be nested in keep <- statement,
  # perhaps to force evaluation of data
  x <- select(data, !!x_name)
  keep <- sapply(lapply(x, unique), length) >= 2
  x_keep <- x_name[keep]
  x_terms <- paste(c("1", x_keep), sep = "", collapse = " + ")
  form <- as.formula(paste(y_name, "~", x_terms))

  # obtain pmm weights
  if (blend > 0) {
    # fit lm model
    fit <- lm(form, data = data, na.action = na.exclude, ...)
    if (exclude_NA) {
      yhat <- fitted(fit, ...)
    } else {
      yhat <- predict(fit, newdata = data, ...)
    }

    # predicted means
    yhat_active <- predict(fit, newdata = active, ...)
    d <- abs(yhat - yhat_active)

    # add noise
    f <- d > 0
    a1 <- ifelse(any(f, na.rm = TRUE),
      min(d[f], na.rm = TRUE), 1
    )
    if (replace) {
      # with replacement
      if (any(!f, na.rm = TRUE)) d <- d + a1
      d <- d^kappa
      l1 <- (1 / d) / (sum(1 / d, na.rm = TRUE))
      l1[is.na(l1)] <- 0 # exclude NA
    } else {
      # no replacement
      if (break_ties) d <- d + runif(length(d), 0, a1 / 10^10)
      nmatch <- min(k, length(d) - 1L) # large nmatch: take all
      if (nmatch == 1L) {
        return(as.integer(data$.row[which.min(d)]))
      }
      d_cut <- sort.int(d, partial = nmatch)[nmatch][1L]
      l1 <- as.integer(data$.row[d <= d_cut & !is.na(d)][1L:nmatch])
    }
  } else {
    # no pmm
    l1 <- 0
  }

  l2 <- 0

  if (replace) {
    l <- (blend * l1) + ((1 - blend) * l2)
    # return matches
    return(sample(1:length(l), size = k, prob = l, replace = TRUE))
  } else {
    # with replacement without blending as of now.
    if (blend < 0.5) {
      return(l2)
    } else {
      return(l1)
    }
  }
}


R/calculate_matches.R
#' Calculates matches for a child by blended distance matching
#'
#' Curve matching is a technology that aims to predict individual growth curves.
#' The method finds persons similar to the target person, and learn the possible
#' future course of growth from the realized curves of the matched individuals.
#'
#' The function finds `k` matches for an individual in the same data set by
#' means of stratified predictive mean matching or nearest neighbour matching.
#'
#' @param data A `data.frame` or `tbl_df`.
#' @param condition Logical expression defining the set of rows in `data`
#'   for which matches will be sought. Missing values are taken as false. If
#'   omitted, all rows will be successively taken as targets. This can result in
#'   intensive computation if `nrow(data)` is large.
#' @param y_name A character vector containing the names of the dependent
#'   variables in `data`.
#' @param x_name A character vector containing the names of predictive variables
#'   in `data` to will go into the linear part of the model.
#' @param e_name A character vector containing the names of the variables for
#'   which the match should be exact.
#' @param t_name A character vector containing the names of the treatment
#'   variables in `data`. The current function will only fit the model to
#'   only the first element of `t_name`.
#' @param subset Logical expression defining the set of rows taken from
#'   `data`. This subset is selected before any other calculations are
#'   made, and this can be used to trim down the size of the data in which
#'   matches are defined and sought.
#' @param k Requested number of matches. The default is `k = 10`.
#' @param replace A logical that indicates whether to match with or without
#'   replacement. The default is `FALSE`.
#' @param blend An integer value between 0 and 1 that indicates the blend
#'   between predictive mean matching with replacement (`1`) and euclidian
#'   distance matching (`0`). The default is `1`.
#' @param kappa A numeric value that serves as the sensitivity parameter for the
#'   inverse distance weighting. Used when drawing with replacement. The default
#'   is `3`.
#' @param break_ties A logical indicating whether ties should broken randomly.
#'   The default (`TRUE`) breaks ties randomly.
#' @param allow_matched_targets A logical that indicates whether the non-active
#'   target cases may be found as a match. The default is `TRUE`.
#' @param include_target A logical that indicates whether the target case is
#'   included in the model. See details. The default is `TRUE`.
#' @param verbose A logical indicating whether diagnostic information should be
#'   printed.
#' @param \dots Arguments passed down to `match_pmm()`.
#' @return An object of class `match_list` which can be post-processed by
#'   the `extract_matches` function to extract the row numbers in
#'   `data` of the matched children. The length of the list will be always
#'   equal to `m` if `replace == TRUE`, but may be shorter if
#'   `replace == FALSE` if the donors are exhausted. The length is zero if
#'   no matches can be found.
#' @author Stef van Buuren 2017
#' @details By default, if the outcome variabe of the target case is observed,
#'   then it used to fit the model, together with the candidate donors. The
#'   default behavior can be changed by setting `include_target = FALSE`.
#'   Note that if `x_name` contains one or more factors, then it is
#'   possible that the factor level of the target case is unique among all
#'   potential donors. In that case, the model can still be fit, but prediction
#'   will fail, and hence no matches will be found.
#'
#'   If `break_ties == FALSE`, the function returns the first `nmatch`
#'   matches as they appear in the order of `data`. This method leads to an
#'   overuse of the first part of the data, and hence underestimates
#'   variability. The better option is to break ties randomly (the default).
#'
#' @references van Buuren, S. (2014). *Curve matching: A data-driven
#'   technique to improve individual prediction of childhood growth*. Annals of
#'   Nutrition & Metabolism, 65(3), 227-233. van Buuren, S. (2012).
#'   *Flexible imputation of missing data*. Boca Raton, FL: Chapman &
#'   Hall/CRC.
#' @examples
#' library("curvematching")
#' data <- datasets::ChickWeight
#' data[543, ]
#'
#' # find matches for observation in row 543 for outcome weight
#' m <- calculate_matches(data, Time == 0 & Chick == 48, y_name = "weight", x_name = c("Time", "Diet"))
#'
#' # row numbers of matched cases
#' extract_matches(m)
#'
#' # data of matched cases
#' data[extract_matches(m), ]
#' @export
calculate_matches <- function(data,
                              condition,
                              y_name,
                              x_name = character(),
                              e_name = character(),
                              t_name = character(),
                              subset = NULL,
                              k = 10,
                              replace = FALSE,
                              blend = 1,
                              break_ties = TRUE,
                              allow_matched_targets = TRUE,
                              include_target = TRUE,
                              kappa = 3,
                              verbose = TRUE, ...) {
  equals_all <- function(x) {
    if (is.null(names(x)) | length(x) == 0) {
      return(character(0))
    }
    cv <- vector("character", length(x))
    for (j in 1:length(x)) {
      xj <- x[[j]]
      if (is.factor(xj)) {
        cv[j] <- paste0(names(x)[j], ' == "', as.character(xj), '"')
      } else {
        cv[j] <- paste0(names(x)[j], " == ", xj)
      }
    }
    paste0(cv, collapse = " & ")
  }

  # check input
  if (!is.data.frame(data)) {
    return(no_match())
  }

  # append unique row id
  data <- data %>%
    mutate(.row = 1:n())

  # restrict to subset if specified
  if (!is.null(subset)) data <- filter(data, {{ subset }})

  # construct target variable
  data <- data %>%
    mutate(.target = {{ condition }}) %>%
    group_by(.data$.target) %>%
    mutate(.seqno = 1:n())

  if (!any(data$.target, na.rm = TRUE)) {
    if (verbose) warning("No target rows conform to condition.")
    return(no_match())
  }

  # no matches if all outcomes in the donor part are missing
  # if (all(is.na(data[!data$.target, y_name]))) return(no_match())
  # FIMXE: add test for absent donor outcomes

  # note: assuming TRUE is last group
  ng <- group_size(data)
  nt <- ng[n_groups(data)]
  data <- ungroup(data)

  # validate variable names
  y_name <- y_name[y_name %in% names(data)]
  x_name <- x_name[x_name %in% names(data)]
  e_name <- e_name[e_name %in% names(data)]
  t_name <- t_name[t_name %in% names(data)]

  # exception handlers
  if (k <= 0 || length(y_name) == 0) {
    return(no_match())
  }

  # select active variables
  sys <- c(".row", ".target", ".seqno")
  var_names <- unique(c(sys, y_name, x_name, e_name, t_name))
  data <- select(data, all_of(var_names))

  # loop over tgts
  l1 <- vector("list", nt)
  names(l1) <- as.vector(unlist(select(filter(data, .data$.target), ".row")))
  for (i in 1:nt) {
    # FIXME: Is it really OK to overwrite data? SvB 20200319
    # define active case
    data <- mutate(data, active = .data$.target & .data$.seqno == i)
    active <- filter(data, active)
    # target_names[i] <- select(active, ".row")

    # re-define candidate set according to allow_matched_targets flag
    if (allow_matched_targets) {
      data <- mutate(data, candidate = !.data$active)
    } else {
      data <- mutate(data, candidate = !.data$.target)
    }

    # trim candidate set by requiring exact matches on
    # variables listed in `e_name`
    trimmed <- select(active, !!e_name)
    cond <- equals_all(trimmed)
    if (length(cond) > 0) {
      cond <- paste0("candidate == TRUE & ", cond)
      expr <- parse(text = cond)
      data <- dplyr::mutate(data, candidate = eval(expr))
    }

    # loop over outcome names
    ny <- length(y_name)
    l2 <- vector("list", ny)
    names(l2) <- y_name
    for (iy in 1:ny) {
      yvar <- y_name[iy]

      # extract subset of candidates
      xy <- filter(data, .data$candidate | .data$active)
      if (!include_target) xy[xy$active, yvar] <- NA

      # split by treatment variables
      if (length(t_name) > 0) {
        # duplicate one tgt row for each treatment level
        # FIXME: as.name will only split on first variable name
        t_name <- t_name[[1]]
        # t_name_list <- as.list(t_name)
        t_unique <- unique(xy[, t_name])
        mutate_call <- quo(!!sym(t_name))
        augment <- slice(active, rep(1:n(), each = nrow(t_unique)))
        augment[, t_name] <- t_unique
        matched <- augment %>%
          bind_rows(filter(xy, .data$candidate)) %>%
          group_by(!!mutate_call) %>%
          do(.row = match_bdm(.,
            y_name = yvar, x_name = x_name, k = k, replace = replace,
            blend = blend, break_ties = break_ties, kappa = kappa, ...
          )) %>%
          mutate(.by = TRUE)
      } else {
        row <- match_bdm(xy,
          y_name = yvar, x_name = x_name, k = k, replace = replace,
          blend = blend, break_ties = break_ties, kappa = kappa, ...
        )
        matched <- tibble(.by = FALSE, .row = list(row))
      }
      l2[[iy]] <- matched
    }
    l1[[i]] <- l2
  }
  # names(l1) <- target_names
  class(l1) <- "match_list"
  l1
}

# Blended distance matching
match_bdm <- function(data, y_name, x_name, k, replace = TRUE, blend = 1,
                      break_ties = TRUE, exclude_NA = FALSE, kappa = 3, ...) {
  if (nrow(data) <= 1) {
    return(no_match())
  }

  if (sum(data$active) > 1) stop("Too many active cases: ", sum(data$active))
  if (sum(data$active) == 0) stop("No active case found.")
  if (blend > 1 | blend < 0) stop("blend needs to be a value between 0 and 1.")

  # keep only independent variables taking at least two values
  # note: apparently, next statement cannot be nested in keep <- statement,
  # perhaps to force evaluation of data
  x <- select(data, !!x_name)
  keep <- sapply(lapply(x, unique), length) >= 2
  x_keep <- x_name[keep]
  x_terms <- paste(c("1", x_keep), sep = "", collapse = " + ")
  form <- as.formula(paste(y_name, "~", x_terms))
  # obtain pmm weights
  if (blend > 0) {
    # fit lm model
    fit <- lm(form, data = data, na.action = na.exclude, ...)
    if (exclude_NA) {
      yhat <- fitted(fit, ...)
    } else {
      yhat <- predict(fit, newdata = data, ...)
    }
    # predicted means
    data <- mutate(data, yhat = yhat)
    yhat_active <- yhat[data$active]
    d <- abs(yhat - yhat_active)
    d[data$active] <- NA

    # add noise
    f <- d > 0
    a1 <- ifelse(any(f, na.rm = TRUE),
      min(d[f], na.rm = TRUE), 1
    )
    if (replace) {
      # with replacement
      if (any(!f, na.rm = TRUE)) d <- d + a1
      d <- d^kappa
      l1 <- (1 / d) / (sum(1 / d, na.rm = TRUE))
      l1[is.na(l1)] <- 0 # exclude NA
    } else {
      # no replacement
      if (break_ties) d <- d + runif(length(d), 0, a1 / 10^10)
      nmatch <- min(k, length(d) - 1L) # large nmatch: take all
      if (nmatch == 1L) {
        return(as.integer(data$.row[which.min(d)]))
      }
      d_cut <- sort.int(d, partial = nmatch)[nmatch][1L]
      l1 <- as.integer(data$.row[d <= d_cut & !is.na(d)][1L:nmatch])
    }
  } else {
    # no pmm
    l1 <- 0
  }

  # obtain distance weights
  if (blend < 1) {
    xdata <- data[, x_keep]
    # transform for euclidian distance calculations
    xdata <- mutate_if(xdata, is.character, as.numeric)
    xdata <- mutate_if(xdata, is.factor, as.numeric)
    xdata <- mutate_if(xdata, is.numeric, function(x) (x - mean(x, na.rm = TRUE)) / sd(x, na.rm = TRUE))
    if (length(x_keep) < 1) xdata <- data.frame(rep(1, nrow(data)))

    # neirest neighbours
    d <- rowSums(sqrt((xdata - c(xdata[data$active, ]))^2))
    d[data$active] <- NA
    # add noise
    f <- d > 0
    a1 <- ifelse(any(f, na.rm = TRUE),
      min(d[f], na.rm = TRUE), 1
    )
    if (replace) {
      # with replacement
      if (any(!f, na.rm = TRUE)) d <- d + a1
      d <- d^kappa
      l2 <- (1 / d) / (sum(1 / d, na.rm = TRUE))
      l2[is.na(l2)] <- 0 # exclude NA
    } else {
      # no replacement
      if (break_ties) d <- d + runif(length(d), 0, a1 / 10^10)
      nmatch <- min(k, length(d) - 1L) # large nmatch: take all
      if (nmatch == 1L) {
        return(as.integer(data$.row[which.min(d)]))
      }
      data$d <- d
      d_cut <- sort.int(d, partial = nmatch)[nmatch][1L]
      l2 <- as.integer(data$.row[d <= d_cut & !is.na(d)][1L:nmatch])
    }
  } else {
    # no nn
    l2 <- 0
  }

  if (replace) {
    l <- (blend * l1) + ((1 - blend) * l2)
    # return matches
    return(sample(1:length(l), size = k, prob = l, replace = TRUE))
  } else {
    # with replacement without blending as of now.
    if (blend < 0.5) {
      return(l2)
    } else {
      return(l1)
    }
  }
}

# Match predictive mean matching (legacy)
match_pmm <- function(data, y_name, x_name, k, break_ties = TRUE,
                      exclude_NA = FALSE, ...) {
  if (nrow(data) <= 1) {
    return(no_match())
  }

  if (sum(data$active) > 1) stop("Too many active cases: ", sum(data$active))
  if (sum(data$active) == 0) stop("No active case found.")

  # keep only independent variables taking at least two values
  # note: apparently, next statement cannot be nested in keep <- statement,
  # perhaps to force evaluation of data
  x <- select(data, !!x_name)
  keep <- sapply(lapply(x, unique), length) >= 2
  x_keep <- x_name[keep]
  x_terms <- paste(c("1", x_keep), sep = "", collapse = " + ")
  form <- as.formula(paste(y_name, "~", x_terms))

  # fit model, and estimate prediction
  fit <- lm(form, data = data, na.action = na.exclude, ...)
  if (exclude_NA) {
    yhat <- fitted(fit, ...)
  } else {
    yhat <- predict(fit, newdata = data, ...)
  }
  data <- mutate(data, yhat = yhat)
  yhat_active <- yhat[data$active]

  d <- abs(yhat - yhat_active)
  d[data$active] <- NA

  # add little noise to break ties
  f <- d > 0
  a1 <- ifelse(any(f, na.rm = TRUE),
    min(d[f], na.rm = TRUE), 1
  )
  if (break_ties) d <- d + runif(length(d), 0, a1 / 10^10)

  nmatch <- min(k, length(d) - 1L) # large nmatch: take all
  if (nmatch == 1L) {
    return(as.integer(data$.row[which.min(d)]))
  }
  d_cut <- sort.int(d, partial = nmatch)[nmatch][1L]
  as.integer(data$.row[d <= d_cut & !is.na(d)][1L:nmatch])
}

#' Returns empty match_list object
#'
#' The empty `match_list` signals that no matches have been found. The
#' empty list should be returned if an error occurs in
#' `calculate_matches()` so that its output remains consistent.
#' @param mode By default a `"list"`
#' @return An empty object of class `match_list`
#' @export
no_match <- function(mode = "list") {
  z <- vector(mode, length = 0)
  class(z) <- "match_list"
  z
}
